probe/endpoint/ebpf.go

```diff
@@ -287,10 +314,45 @@ func (t *EbpfTracker) isDead() bool {
 
 func (t *EbpfTracker) stop() {
 	t.Lock()
-	alreadyDead := t.dead
-	t.dead = true
+	alreadyDead := t.dead || t.stopping
+	t.stopping = true
 	t.Unlock()
-	if !alreadyDead && t.tracer != nil {
-		t.tracer.Stop()
+
+	// Do not call tracer.Stop() in this thread, otherwise tracer.Stop() will
+	// deadlock waiting for this thread to pick up the next event.
```

Comment:
BTW, the `stop` channel should have a higher priority than the other channels here: https://github.com/weaveworks/tcptracer-bpf/blob/master/pkg/tracer/tracer.go#L80-L99